### DISCUSSION TITLE
Performance improvement for the SDK and browser-plugin

### DISCUSSION THREAD

**Discussion by kcramakrishna** [Digital-Assistant/Digital_Assistant_Client#369] [General]
Areas of exploration for improving performance:
1. SDK: We can optimise delivery/download times by serving the code from S3 instead of github.  With S3, we can enable compression as well as browser caching.
2. Browser plugin: We need to see how to optimise js code to improve performance. - Explore and Identify performance bottlenecks, otmise code, also optimise JIT JS execution instead of executing everything at the beginning. 

https://web.dev/codelab-text-compression/ (gzip in header)
https://www.section.io/engineering-education/understanding-browser-caching/

https://www.w3schools.com/js/js_performance.asp
https://www.codemotion.com/magazine/frontend/javascript/javascript-speed-up-tips/

**Comment by kcramakrishna** [comment]
Did a google for speeding up attaching eventListeners. Found some promising articles: 

1. https://makandracards.com/makandra/657-efficiently-add-an-event-listener-to-many-elements
2. https://ehsangazar.com/optimizing-javascript-event-listeners-for-performance-e28406ad406c  (See the "Source" at the end of this article) 
3. https://www.digitalocean.com/community/tutorials/js-speed-up-scroll-events
4. https://stackoverflow.com/questions/41798755/event-listeners-efficiency  (Should we be using 'onClick' instead of listener ?) 
5. Or "passive event listeners" - https://web.dev/uses-passive-event-listeners/

**Reply by yureshwar** [reply]
First link seems to be more interesting but as when i have gone through the documentation of Addeventlistner there are two things that we need to consider.

1. event.stopPropagation - which stops the bubbling
2. event.stopImmediatePropagation - which also stops bubbling and also other eventListners on the DomNode

https://javascript.info/bubbling-and-capturing

There are options available for addeventlistner function which can be passed those are capture, passive, once, signal

1. passive is mainly improves the performance for the handlers that are linked with scrolling effect
2. capture which makes the functionality invoked before the bubbling starts with top down approach where bubbling uses bottom up approach
3. signal will help in aborting the event getting executed
4. once will run the event only one time.

https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener

**Reply by yureshwar** [reply]
We are not going to go through the first link approach as we may encounter issue in recording due to stopPropagation in which the recording may not happen.

https://stackoverflow.com/questions/4616694/what-is-event-bubbling-and-capturing

Hence we are going to attach the event listeners only during recording time only.

**Comment by kcramakrishna** [comment]
**Other ways to speed up**:
There is a concept called worklets in browser. This may help speed up event listeners.
https://developer.mozilla.org/en-US/docs/Web/API/Worklet 

Also webAssembly: This may be a good option.
https://developer.mozilla.org/en-US/docs/WebAssembly


**Multithreading in javascript** ?

https://www.oreilly.com/library/view/multithreaded-javascript/9781098104429/
https://web.dev/webassembly-threads/

C/C++ libraries can be included in Node but NOT in browser plugin:
https://qr.ae/pvjDX8

Using C/C++ libraries in javascript - general standalone apps but nor browser plugin
https://rapidjson.org/md_doc_dom.html
https://medium.com/jspoint/a-simple-guide-to-load-c-c-code-into-node-js-javascript-applications-3fcccf54fd32
https://www.quora.com/Is-it-possible-to-include-a-C++-library-in-a-JavaScript-application

**Comment by kcramakrishna** [comment]
Cost of hosting on CDN vs. hosting on our server: https://www.freecodecamp.org/news/do-not-use-s3-for-static-assets/ 
It may make sense to serve it from our webserver instead of CDN (at least for initial months till we see traction) We need to do our own calculation.

**Comment by kcramakrishna** [comment]
Partial list of performance improvements:
1. Compression during SDK download. Previously, the download size was a few MB. Right now, the size is a few hundred KB?
2. Enabled Browser caching: Download SDK only once for every user. Older link used to download SDK for every page load which is expensive in terms of bandwidth as well as time taken. Compression in previous step combined with caching reduces time multifold.
3. Optimisation in adding listeners: Previously, we were adding listeners to every clickable event during page load. This is CPU intensive. This was a bit wasteful and slows down the page load on systems with older CPUs or weaker CPUs. The current approach adds listeners during 'record' mode only. This save significant CPU usage but not much difference in load times.
4. More improvements to be listed here.


Other improvements:
1. Went from javascript to ReactJS + AntD + Typescript + SCSS. Code is super modular. Easier to fix bugs and make feature changes.
2. Used many standard frameworks as best practices - logging, error-reporting, error handling etc.
3. Moving to Shadow-DOM - this will eliminate need to test for individual sites. Our CSS will never interfere with applications and vice-versa.
